### PR TITLE
Problem: Image Owner cant update visibility/access for others

### PR DESCRIPTION
**Pull request by steve-gregory**
## Solution: CRUD support for ImageVersionMembership
- Create new tasks for long-running 'add' and 'remove' membership
- Remove 'membership' as a writeable field for ImageVersion
- Using POST/DELETE of /v2/image_version_memberships instead for
a better experience
- Call tasks in POST/DELETE to avoid unnecessary delays.
- If user is calling /v2/groups?search=... provide all group names
- [x] Updated image statistics
- [x] Email address patterns to grant future-user access without changing ImageVersionMembership.
- [x] Update `monitor_machines_for` to ensure that `Application.access_list` is respected for new and existing users.

## Checklist before merging Pull Requests
- [ ] New test(s) included to reproduce the bug/verify the feature
- [ ] Documentation created/updated at [Example link to documentation](https://example.test/doc#new_section) to give context to the feature
- [ ] Reviewed and approved by at least one other contributor.
- [ ] [Reviewed and approved Troposphere companion PR](https://github.com/cyverse/troposphere/pull/706)
- [ ] [Reviewed and approved Clank PR(MERGE LAST!)](https://github.com/cyverse/clank/pull/221)

### CODE REVIEW

**Comment by julianpistorius** [comment]
Might have to talk me through this one. Looks interesting.

**Comment by julianpistorius** [comment]
Also test failures don't seem spurious.

File: service/tasks/monitoring.py [hunk 1]
```diff
@@ -286,22 +286,35 @@ def update_image_membership(account_driver, cloud_machine, db_machine):
...
         identifier = has_machine_request.new_machine.identifier
         main_account_driver = get_account_driver(provider)
+        #Extend to include based on information in the machine request
         shared_projects_from_main =  main_account_driver.shared_images_for(identifier, None)
         shared_group_names.extend(p.name for p in shared_projects_from_main if p)
 
+
+    #Extend to include new names found by application pattern_match
+    parent_app = db_machine.application_version.application
+    matching_users = list(parent_app.get_users_from_access_list().values_list('username', flat=True))
```
Comment (julianpistorius) [inline service/tasks/monitoring.py]: TODO: Julian have a look why it's slow.

File: service/tasks/monitoring.py [hunk 2]
```diff
@@ -286,22 +286,35 @@ def update_image_membership(account_driver, cloud_machine, db_machine):
     """
     Given a cloud_machine and db_machine, create any relationships possible for ProviderMachineMembership and ApplicationVersionMembership
     """
+    import ipdb;ipdb.set_trace()
```
Comment (julianpistorius) [inline service/tasks/monitoring.py]: :)

File: core/models/pattern_match.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+class PatternMatch(models.Model):
+
+    """
+    pattern - the actual string to be matched on
+    type - logic to use for matching the string
+    """
+    pattern = models.CharField(max_length=256)
+    type = models.ForeignKey(MatchType)
+    created_by = models.ForeignKey(AtmosphereUser)
```
Comment (steve-gregory) [inline core/models/pattern_match.py]: Todo: add `exclusion` or `negate` option to "blacklist nefarious users".

**Comment by julianpistorius** [comment]
Looks good. 

Will look at performance issue & black listing option.

**Comment by steve-gregory** [comment]
ToDo: Re-order migrations.

**Comment by julianpistorius** [comment]
Failure in that test is due to summarized metrics. See my PR-PR:

https://github.com/steve-gregory/atmosphere/pull/13/files#diff-c7e2d73d772596cb73af5675c93908ccR181

**Comment by steve-gregory** [comment]
Performance and black-listing issues addressed. This final commit should pass all tests.

**Comment by steve-gregory** [comment]
🚅 :shipit: 🚄

**Pull request merged**